src/core/layout/qgslayoutexporter.cpp

```diff
@@ -43,26 +57,693 @@ void QgsLayoutExporter::renderPage( QPainter *painter, int page )
+  mapItemPos.rx() -= exportRegion.left();
+  mapItemPos.ry() -= exportRegion.top();
+
+  double xmin = mapExtent.xMinimum() - mapItemPos.x() * xRatio;
+  double ymax = mapExtent.yMaximum() + mapItemPos.y() * yRatio;
+  QgsRectangle paperExtent( xmin, ymax - destinationHeight * yRatio, xmin + destinationWidth * xRatio, ymax );
+
+  double X0 = paperExtent.xMinimum();
+  double Y0 = paperExtent.yMinimum();
+
+  if ( dpi < 0 )
+    dpi = mLayout->context().dpi();
+
+  int widthPx = static_cast< int >( dpi * destinationWidth / 25.4 );
+  int heightPx = static_cast< int >( dpi * destinationHeight / 25.4 );
+
+  double Ww = paperExtent.width() / widthPx;
+  double Hh = paperExtent.height() / heightPx;
+
+  // scaling matrix
+  double s[6];
+  s[0] = Ww;
+  s[1] = 0;
+  s[2] = X0;
+  s[3] = 0;
+  s[4] = -Hh;
+  s[5] = Y0 + paperExtent.height();
+
+  // rotation matrix
+  double r[6];
+  r[0] = std::cos( alpha );
+  r[1] = -std::sin( alpha );
+  r[2] = xCenter * ( 1 - std::cos( alpha ) ) + yCenter * std::sin( alpha );
+  r[3] = std::sin( alpha );
+  r[4] = std::cos( alpha );
+  r[5] = - xCenter * std::sin( alpha ) + yCenter * ( 1 - std::cos( alpha ) );
+
+  // result = rotation x scaling = rotation(scaling(X))
+  a = r[0] * s[0] + r[1] * s[3];
+  b = r[0] * s[1] + r[1] * s[4];
+  c = r[0] * s[2] + r[1] * s[5] + r[2];
+  d = r[3] * s[0] + r[4] * s[3];
+  e = r[3] * s[1] + r[4] * s[4];
+  f = r[3] * s[2] + r[4] * s[5] + r[5];
+}
+
+QImage QgsLayoutExporter::createImage( const QgsLayoutExporter::ImageExportSettings &settings, int page, QRectF &bounds, bool &skipPage ) const
+{
+  bounds = QRectF();
+  skipPage = false;
+
+  if ( settings.cropToContents )
+  {
+    if ( mLayout->pageCollection()->pageCount() == 1 )
+    {
+      // single page, so include everything
+      bounds = mLayout->layoutBounds( true );
+    }
+    else
+    {
+      // multi page, so just clip to items on current page
+      bounds = mLayout->pageItemBounds( page, true );
+    }
+    if ( bounds.width() <= 0 || bounds.height() <= 0 )
+    {
+      //invalid size, skip page
+      skipPage = true;
+      return QImage();
+    }
+
+    double pixelToLayoutUnits = mLayout->convertToLayoutUnits( QgsLayoutMeasurement( 1, QgsUnitTypes::LayoutPixels ) );
+    bounds = bounds.adjusted( -settings.cropMargins.left() * pixelToLayoutUnits,
+                              -settings.cropMargins.top() * pixelToLayoutUnits,
+                              settings.cropMargins.right() * pixelToLayoutUnits,
+                              settings.cropMargins.bottom() * pixelToLayoutUnits );
+    return renderRegionToImage( bounds, QSize(), settings.dpi );
+  }
+  else
+  {
+    return renderPageToImage( page, settings.imageSize, settings.dpi );
+  }
+}
+
+QString QgsLayoutExporter::generateFileName( const PageExportDetails &details ) const
+{
+  if ( details.page == 0 )
+  {
+    return details.directory + '/' + details.baseName + '.' + details.extension;
```

Comment:
A ` QStringLiteral().arg()` is not to be preferred in this case?

Reply:
Not sure here - my understanding is that the QString single character based methods are very fast, so I suspect that the extra allocations here would be faster than parsing the string literal for the % tokens.

Reply:
Probably QStringBuilder would be preferable here actually

Reply:
> Probably QStringBuilder would be preferable here actually

Good point